operator/src/test/java/org/keycloak/operator/testsuite/apiserver/ApiServerHelper.java

```diff
@@ -0,0 +1,85 @@
+
+import org.keycloak.operator.Utils;
+
+import io.fabric8.kubeapitest.KubeAPIServer;
+import io.fabric8.kubernetes.api.model.apps.StatefulSet;
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientBuilder;
+import io.fabric8.kubernetes.client.informers.ResourceEventHandler;
+
+public class ApiServerHelper {
+
+    private KubeAPIServer kubeApi;
+
+    public ApiServerHelper() {
+        kubeApi = new KubeAPIServer();
+        kubeApi.start();
+    }
+
+    public void stop() {
+        kubeApi.stop();
+    }
+
+    public KubernetesClient createClient(String namespace) {
+        Config config = Config.fromKubeconfig(kubeApi.getKubeConfigYaml());
+        config.setNamespace(namespace);
+        var result = new KubernetesClientBuilder().withConfig(config).build();
+
+        // fake statefulset controller - we don't have to worry about closing this
+        // that will happen automatically when the client is closed
+        result.apps().statefulSets().inAnyNamespace().inform(new ResourceEventHandler<StatefulSet>() {
+
+            @Override
+            public void onAdd(StatefulSet obj) {
+                updateStatefulSet(obj);
+            }
+
+            private void updateStatefulSet(StatefulSet obj) {
+                int replicas = obj.getSpec().getReplicas();
+                String revision = Utils.hash(List.of(obj.getSpec())).substring(1);
+                if (!Objects.equals(Optional.ofNullable(obj.getStatus().getReplicas()).orElse(0), replicas)
+                        || !Objects.equals(revision, obj.getStatus().getUpdateRevision())
+                        || !Objects.equals(obj.getStatus().getCurrentRevision(), obj.getStatus().getUpdateRevision())) {
+                    // generate intermediate rolling events
+                    int actualReplicas;
+                    if (!Objects.equals(revision, obj.getStatus().getUpdateRevision())) {
+                        actualReplicas = 0;
+                    } else if (replicas == 0) {
+                        actualReplicas = Math.max(replicas, Optional.ofNullable(obj.getStatus().getReplicas()).orElse(0) - 1);
+                    } else {
+                        actualReplicas = Math.min(replicas, Optional.ofNullable(obj.getStatus().getReplicas()).orElse(0) + 1);
+                    }
+                    obj = result.getKubernetesSerialization().clone(obj);
+                    obj.getStatus().setReplicas(actualReplicas);
+                    obj.getStatus().setReadyReplicas(actualReplicas);
+                    obj.getStatus().setUpdateRevision(revision);
+                    if (actualReplicas == replicas) {
+                        obj.getStatus().setCurrentRevision(revision);
+                    }
+                    obj.getMetadata().setResourceVersion(null);
+                    result.resource(obj).updateStatus();
+                }
+            }
```

Comment:
I found this a bit hard to read. I tried to clean it up it https://github.com/shawkins/keycloak/pull/17. Additionally, I changed the logic for rolling update to better mimic actual K8s behaviour (that updates are propagated one Pod at a time). Let me know what you think.

Reply:
> that updates are propagated one Pod at a time

Your pr doesn't fully do that. When it decrements the status replicas, it's still setting the updateRevision, so the next iteration we'll just immediately scale back up to the full replica count. You'll need more complex logic to achieve what you are describing - differentiating between current and ready replicas. 

I didn't consider that necessary because our operator only looks at ready replicas.

Reply:
Yes, it doesn't fully mimic the actual behaviour – it doesn't roll all Pods but only one. Still it felt more accurate to me that there's a "1 pod down" spike. I wouldn't want to go beyond that – as you said, the Operator doesn't care much what Pods got updated. But I can pull back that change and keep scaling down to 0, if you want.

Reply:
Good point, updated the PR